ra/ra.go

```diff
@@ -2417,22 +2415,16 @@ func (ra *RegistrationAuthorityImpl) NewOrder(ctx context.Context, req *rapb.New
 			authzAge = (profile.pendingAuthzLifetime - authz.Expires.Sub(ra.clk.Now())).Seconds()
 		}
 
-		// If the identifier is a wildcard and the existing authz only has one
-		// DNS-01 type challenge we can reuse it. In theory we will
-		// never get back an authorization for a domain with a wildcard prefix
-		// that doesn't meet this criteria from SA.GetAuthorizations but we verify
-		// again to be safe.
-		if strings.HasPrefix(name, "*.") &&
-			len(authz.Challenges) == 1 && authz.Challenges[0].Type == core.ChallengeTypeDNS01 {
-			authzID, err := strconv.ParseInt(authz.ID, 10, 64)
-			if err != nil {
-				return nil, err
-			}
-			newOrderAuthzs = append(newOrderAuthzs, authzID)
-			ra.authzAges.WithLabelValues("NewOrder", string(authz.Status)).Observe(authzAge)
-			continue
-		} else if !strings.HasPrefix(name, "*.") {
-			// If the identifier isn't a wildcard, we can reuse any authz
+		// If the identifier isn't a DNS name, or isn't a wildcard, we can
+		// reuse any authz.
+		//
+		// If the identifier is a wildcard DNS name and the existing authz only
+		// has one DNS-01 type challenge we can reuse it. In theory we will
+		// never get back an authorization for a DNS name with a wildcard prefix
+		// that doesn't meet this criteria from SA.GetAuthorizations but we
+		// verify again to be safe.
+		if ident.Type != identifier.TypeDNS || !strings.HasPrefix(ident.Value, "*.") ||
+			(len(authz.Challenges) == 1 && authz.Challenges[0].Type == core.ChallengeTypeDNS01) {
```

Comment:
Nice work combining these two stanzas. Looking at them again I'm wondering why we have this fallthrough logic. It seems like instead we should error when the expected conditions are not met:


```
if ident.Type == identifier.TypeDNS && strings.HasPrefix(ident.Value, "*.")
  && (len(authz.Challenges) != 1 || authz.Challenges[0].Type != core.ChallengeTypeDNS01) {
  return nil, errors.New("authorization reuse for wildcard DNS name yielded invalid authorization object")
}

// Parse the authz ID, append, etc

// No need for the final stanza of deleting from the map and appending to the missing list.
```